feat: RPC modernization APIs

jennifer-richards opened the pull request.

jennifer-richards committed:
feat: Add rpc_person API call

jennifer-richards committed:
chore: Add OpenAPI yaml for RPC API

jennifer-richards committed:
Merge pull request #6388 from ietf-tools/main

chore: Merge main branch

rjsparks committed:
feat: api endpoint for docs submitted to the rpc

rjsparks committed:
chore: remove some debug

jennifer-richards committed:
Merge pull request #6390 from rjsparks/submitted_to_rpc

feat: api endpoint for docs submitted to the rpc

rjsparks committed:
feat: api for created demo people

rjsparks committed:
feat: optimize fetching persons

jennifer-richards committed:
Merge pull request #6391 from rjsparks/persons_optimization

feat: optimize fetching persons

rjsparks committed:
feat: api for creating demo drafts (#6402)

jennifer-richards committed:
fix: Typo in rpcapi.yaml

jennifer-richards committed:
refactor: Allow existing Person in create_demo_person (#6398)

rjsparks committed:
fix: include pubreq docevents in demo drafts (#6404)

jennifer-richards committed:
fix: Minor API fixes (#6405)

jennifer-richards committed:
chore: Document 404 from rpcapi get_person_by_id

rjsparks committed:
feat: adding rev to demo doc creation (#6425)

* feat: adding rev to demo doc creation

* fix: remove attempt to control required

jennifer-richards committed:
refactor: Replace api token checks with decorator (#6434)

* feat: Add @requires_api_token decorator

* refactor: Use @requires_api_token

* refactor: Tweak api token endpoints

This might be drifting from the design intent, but at least
uses the defined endpoint values. Further cleanup may well
be needed.

jennifer-richards committed:
refactor: Improve usability of @requires_api_token

jennifer-richards committed:
feat: get_draft_by_id api call (#6446)

rjsparks committed:
fix: construct the cdn photo url correctly (#6491)

rjsparks committed:
chore: restructure the rpc api (#6505)

rjsparks committed:
fix: authenticate person api correctly

jennifer-richards committed:
chore: Fix plain_name lookup

jennifer-richards committed:
Merge pull request #6551 from ietf-tools/main

chore: Merge main into feat/rpc-api

jennifer-richards committed:
feat: subject_id -> Person api call (#6566)

* feat: subject_id -> Person api call

* doc: Add error responses to openapi spec

jennifer-richards committed:
Merge pull request #6631 from ietf-tools/main

chore: Merge main into feat/rpc-api

jennifer-richards committed:
Merge pull request #6717 from ietf-tools/main

chore: Merge main into feat/rpc-api

jennifer-richards committed:
Merge remote-tracking branch 'upstream/main' into feat/rpc-api

# Conflicts:
#	ietf/doc/models.py

jennifer-richards committed:
Merge pull request #6788 from ietf-tools/main

chore: Merge main into feat/rpc-api

jennifer-richards committed:
Merge branch 'main' into feat/rpc-api

# Conflicts:
#	ietf/doc/models.py

jennifer-richards committed:
Merge pull request #6815 from ietf-tools/main

chore: Merge main into feat/rpc-api

rjsparks committed:
feat: rpc api drafts by names (#6853)

rjsparks committed:
Merge branch 'main' into feat/rpc-api

rjsparks committed:
ci: Merge branch 'main' into feat/rpc-api

rjsparks committed:
feat: get stream rfcs were first published into (#7814)

* feat: get stream rfcs were first published into

* chore: black

* fix: deal with the case of no DocHistory having a stream

rjsparks committed:
Merge branch 'main' into feat/rpc-api

rjsparks committed:
fix: return "none" from drafts_by_names if a draft has a stream of None (#7863)

jennifer-richards committed:
feat: API changes needed for "real" draft imports (#7877)

* style: black

* fix: submitted as DateTime, not Date

* fix: pk -> id in API

* feat: rpc_draft source_format

* feat: add shepherd to rpc_draft() api

* fix: "unknown" src fmt instead of error

* feat: add intended_std_level to api

* refactor: blank, not None, for shepherd / std_level

* style: black

jennifer-richards committed:
fix: typo in drafts_by_names() (#7912)

jennifer-richards committed:
Merge branch 'main' into feat/rpc-api

# Conflicts:
#	dev/deploy-to-container/settings_local.py

rjsparks committed:
Merge branch 'main' into feat/rpc-api

jennifer-richards committed:
Merge pull request #8071 from ietf-tools/main

chore: merge main into feat/rpc-api

rjsparks committed:
feat: support for building rfc authors (#7983)

* feat: support for building rfc authors

* chore: copyrights

jennifer-richards committed:
Merge pull request #8124 from ietf-tools/main

chore: merge main into feat/rpc-api

rjsparks committed:
feat: api to gather draft authors (#8126)

* feat: api to gather draft authors

* chore: black

jennifer-richards committed:
Merge pull request #8257 from ietf-tools/main

chore: merge main to feat/rpc-api

jennifer-richards committed:
ci: tag feature branch release images

jennifer-richards committed:
chore: fix git nonsense

jennifer-richards committed:
Merge branch 'main' into feat/rpc-api

jennifer-richards committed:
Merge pull request #8281 from ietf-tools/main

chore: merge main to feat/rpc-api

jennifer-richards committed:
feat: API for RFC metadata fetch/filtering (#8291)

* feat: RFC API for rfceditor website (WIP)

* feat: pagination + ordered RFC index

* feat: filter by publication date

* feat: stream + stream filtering

* feat: DOI

* feat: group/area for RFCs

* feat: group/area filtering

* feat: result sorting

* refactor: send rfc number, not name

* feat: search rfc title/abstract

* style: Black

* feat: add 'status' field

* feat: filter by 'status'

* style: remove redundant parentheses

* feat: add updated_by/obsoleted_by fields

holloway committed:
feat: add 'abstract' to rpc api

jennifer-richards committed:
Merge pull request #8326 from holloway/rpc-api-abstract

feat: add 'abstract' to rpc api

jennifer-richards committed:
Merge pull request #8327 from ietf-tools/main

chore: merge main into feat/rpc-api

jennifer-richards committed:
chore: fix unused/duplicate imports

jennifer-richards committed:
chore: fix mypy lint

jennifer-richards committed:
chore: unused import

jennifer-richards committed:
feat: retrieve single rfc, including text (#8346)

* feat: retrieve single rfc

Use RFC number instead of doc PK as id

* feat: include text in single-rfc response

* chore: drop doc id from api response

jennifer-richards committed:
fix: many=True for identifiers (#8425)

jennifer-richards committed:
Merge branch 'main' into feat/rpc-api

jennifer-richards committed:
feat: add more rfc api fields (many stubs)

jennifer-richards committed:
Merge pull request #8492 from jennifer-richards/more-fields

feat: add more rfc api fields (many stubs)

holloway committed:
chore: adding postscript (ps) to rfc meta serializer (#8560)

jennifer-richards committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into feat/rpc-api

# Conflicts:
#	ietf/doc/models.py

jennifer-richards committed:
Merge pull request #8772 from ietf-tools/main

chore: merge main into feat/rpc-api

jennifer-richards committed:
chore: merge main into feat/rpc-api

rjsparks committed:
ci: merge branch 'main' into feat/rpc-api

jennifer-richards committed:
chore: merge main into feat/rpc-api

jennifer-richards committed:
fix: acknowledge not-issued in RfcStatusSlugT

jennifer-richards committed:
Merge pull request #8914 from jennifer-richards/not-issued-lint

fix: acknowledge not-issued in RfcStatusSlugT

kesara committed:
feat: Add API call to get references

kesara committed:
fix: Filter drafts

kesara committed:
chore: Optimize the data query

kesara committed:
test: Test for norminative references API call

kesara committed:
chore: Fix typos in tests

kesara committed:
chore: Fix typos

jennifer-richards committed:
Merge pull request #8918 from kesara/feat/refs

feat: Add API call to get normative references

kesara committed:
refactor: Separate demo logic (#8937)

* refactor: Separate demo logic

* chore: Skip tests

rjsparks committed:
fix: trailing slashes for all rpc api endpoints (#8940)

kesara committed:
chore: Add RPC references API call to OpenAPI spec (#8941)

* chore: Remove line noise

* chore: Add RPC references API call to OpenAPI spec

* chore: Update rpcapi.yaml

Co-authored-by: Robert Sparks <rjsparks@nostrum.com>

---------

Co-authored-by: Robert Sparks <rjsparks@nostrum.com>

kesara committed:
fix: Fix OpenAPI spec errors (#8943)

* chore: Remove more line noise

* fix: Fix OpenAPI spec errors

jennifer-richards committed:
feat: include picture URL in rpc_person API (#9009)

jennifer-richards committed:
Merge branch 'main' into feat/rpc-api

# Conflicts:
#	docker/configs/settings_local.py

jennifer-richards committed:
feat: person search endpoint (#9062)

* feat: person search endpoint

* refactor: address review comments

  * improved naming of operation/components in API schema
  * reused Person schema component
  * added serializers_rpc.py

* chore: RpcPersonSerializer -> PersonSerializer

Better matches the hand-written schema.

* fix: search for entire term, not word-by-word

* fix: only look at name/plain in search

Including ascii / ascii_short might be useful eventually, but since we only show plain_name in the response it can cause confusing results.

By the same reasoning we could remove email__address as well, but that's useful and I expect we'll include email addresses in our response soon anyway.

jennifer-richards committed:
Merge pull request #9090 from ietf-tools/main

chore: merge main into feat/rpc-api

jennifer-richards committed:
Merge pull request #9108 from ietf-tools/main

chore: merge main into feat/rpc-api

jennifer-richards committed:
refactor: reimplement purple API in django-rest-framework (#9097)

* refactor: rpc_person -> PersonViewSet

* refactor: rpc_subject_person -> SubjectPersonView

* refactor: rpc_persons -> RpcPersonsView

* refactor: move get_persons into PersonViewSet

Changes the interface to return a list of Persons instead of a map from ID to name.

* refactor: rpc_draft -> DraftViewSet

* refactor: drafts_by_names -> DraftsByNameView

* refactor: submitted_to_rpc -> DraftViewSet

* refactor: rfc_original_stream -> RfcViewSet

* refactor: rpc demo APIs -> viewset

* refactor: get_draft_refs -> DraftViewSet

* refactor: persons_by_email -> PersonViewSet

* refactor: rfc_authors -> RfcViewSet

* refactor: draft_authors -> DraftViewSet

* refactor: avoid \x00 in regex validator

Gets turned into a literal nul somewhere in the process of generating a schema and building a Python client for purple. This has the same effect but avoids the nul.

* fix: missing arg on references() action

* style: ruff, remove unused imports

* style: ruff ruff

* chore: remove rpcapi.yaml

* refactor: move API to /api/purple

Side effect is that the purple API client is named PurpleApi instead of RpcApi.

* fix: get_draft_authors returns DraftWithAuthors

* fix: distinguish CharField flavors

* fix: no serializer validators for draft name/title

This prevents at least one existing draft from being looked up.

* fix: get_draft_authors works with str, not int

* Revert "refactor: avoid \x00 in regex validator"

This reverts commit 63f40cf2

* Revert "Revert "refactor: avoid \x00 in regex validator"" (#9111)

This reverts commit d8656f470045c21542824d7b9b9be41bdcb8866d.

jennifer-richards committed:
ci: only migrate blobdb if it is configured

jennifer-richards committed:
feat: add email/url to purple person API (#9127)

jennifer-richards committed:
Merge pull request #9316 from ietf-tools/main

chore: merge main into feat/rpc-api

jennifer-richards committed:
Merge branch 'main' into feat/rpc-api

# Conflicts:
#	requirements.txt

rudimatz committed:
feat: expose consensus in submission api

jennifer-richards committed:
Merge pull request #9504 from rudimatz/feat/add-consensus-submission

feat: expose consensus in submission api

jennifer-richards committed:
feat: subseries api for red (#9556)

* refactor: central def of subseries doc types

* feat: subseries doc API

* refactor: optimize queries via prefetch

Reduced 4500 to 18 queries

* chore: remove debug

* fix: fix serialization of draft field

* refactor: clean up prefetch a bit

* feat: filter by subseries type

* fix: restore max_limit for RFC pagination

jennifer-richards committed:
feat: add subseries+stub titlepage_name to rfc serializer (#9569)

* feat: add subseries to RfcMetadataSerializer

* feat: titlepage_name for RfcAuthorSerializer

Always blank for now

* chore: update copyrights

* refactor: use py3.12 typing syntax

jennifer-richards committed:
chore: merge main to feat/rpc-api

jennifer-richards committed:
fix: renumber migrations

rudimatz committed:
feat: add consensus on FullDraftSerializer

rjsparks commented on dev/deploy-to-container/settings_local.py:
```diff
@@ -71,11 +71,10 @@
 APP_API_TOKENS = {
-  "ietf.api.views.ingest_email_test": ["ingestion-test-token"]
+  "ietf.api.views.ingest_email_test": ["ingestion-test-token"],  # Not a real secret
+  "ietf.api.views_rpc" : ["devtoken"],  # For RPC dev work only, remove before merging to main!!!
```
Replace this with "not a real secret"?

rjsparks commented on docker/configs/settings_local.py:
```diff
@@ -100,3 +100,7 @@
     }
+
+APP_API_TOKENS = {
+    "ietf.api.views_rpc" : ["devtoken"],
```
not a real secret?

rjsparks commented on ietf/api/urls_rpc.py:
```diff
@@ -0,0 +1,33 @@
+]
+
+# add routers at the end so individual routes can steal parts of their address
+# space (specifically, ^person/ routes so far)
```
verify that this is current?

jennifer-richards replied:
The specific example was outdated. This will use a current example:
```suggestion
# space (e.g., ^rfc/publish/ superseding the ^rfc/ routes of RfcViewSet)
```

rjsparks commented on ietf/api/views_rpc_demo.py:
```diff
@@ -0,0 +1,98 @@
+# Copyright The IETF Trust 2023-2025, All Rights Reserved
```
Should this capability come to main? I'm thinking it shouldn't.

rjsparks commented on ietf/doc/models.py:
```diff
@@ -714,14 +724,22 @@ def contains(self):
     def part_of(self):
         return self.related_that("contains")
-
+    
```
whitespace

rjsparks commented on ietf/doc/models.py:
```diff
@@ -714,14 +724,22 @@ def contains(self):
+            return self.docevent_set.filter(type="iesg_approved").order_by("-time").first()
+        elif self.stream_id in ["iab", "irtf", "ise"]:
+            return self.docevent_set.filter(type="requested_publication").order_by("-time").first()
+        #elif self.stream_id == "editorial": #TODO
```
we have one now to look at and then add this branch to the logic

rjsparks reviewed (commented):
First pass inspecting "can this safely come into main".
I haven't done the "are the apis all sufficiently access guarded" read yet.

jennifer-richards committed:
Merge pull request #9605 from rudimatz/feat/add-consensus-submission

feat: add consensus on FullDraftSerializer

rudimatz committed:
feat: add type field in serializer

rudimatz committed:
Merge pull request #9776 from rudimatz/feat/add-type-for-group

feat: add type field in group serializer

rudimatz committed:
feat: tag subseries API endpoints for purple (#9763)

* feat: tag subseries API endpoints for purple

* ruff

rudimatz committed:
feat: change slugs/names (#9778)

* change slugs/names

* change slug names

* fix: update RfcStatusSlugT

jennifer-richards committed:
fix: remove double tag (#9787)

This was meant to include the API call in both purple and red API clients.
It seems this does not work, at least with some generators.
Need to investigate further, but we should be able to work around it.

jennifer-richards committed:
chore: merge main into feat/rpc-api

jennifer-richards committed:
feat: rfc authors (#9937)

* feat: rfc authors

* fix: distinct rfc search results

* fix: include titlepage_name in author name searches

* fix: add is_editor to rfcauthor model. Adjust FK on_delete. Tweak admin.

* fix: renumber migration

* refactor: realistic titlepage_name in factory

* refactor: comment + rename iteration var

* chore: bump copyright year

* chore: remove country from RfcAuthor

Not planning to track this

* refactor: make blank=False explicit

* chore: remove country from admin

* fix: author_list() for type=rfc

* fix: blankable RfcAuthor.person

* feat: RfcAuthor-aware document_json()

* feat: limit docs to rfcs for RfcAuthor admin

* test: document_json authors

* fix: use author_names() for get_document_emails()

* feat: suggest affiliation based on RfcAuthor

* chore: revert "remove country from RfcAuthor"

This reverts commit 3044d10439bed2137e28356f27250600ce7a2529.

* chore: revert "remove country from admin"

This reverts commit 208879368fa5be5ecf79ae67b0be636db8a7e81f.

* feat: use rfcauthors for nomcom eligibility (#9629)

* refactor: author-based eligibility cleanup

* feat: use rfcauthor recs for nomcom eligbility

* chore: remove commented code

* refactor: factor out helper for testing

* test: test_get_qualified_author_queryset

* fix: restore a necessary import

* test: fix test_elig_by_author

* test: fix test_decorate_volunteers_with_qualifications

* test: add comment

* fix: drop test for draft-rfceditor state

Attempted to limit to drafts literally in the queue, but was not a
valid check when looking back in time. As a practical matter, the test
is not necessary.

* fix: exclude double counting, not rfc pub state

* test: update test

---------

Co-authored-by: Robert Sparks <rjsparks@nostrum.com>

jennifer-richards committed:
fix: renumber migrations

jennifer-richards committed:
feat: basic RfcAuthor API (#9940)

* refactor: move get_rfc_authors API URL

* refactor: drop format_suffixes from router

Creates a bunch of API endpoints we have no intention of ever using

* feat: RfcAuthor API (WIP)

* fix: remove debug code

* style: remove stray whitespace

jennifer-richards committed:
fix: authors API for xfer (#9961)

* fix: partial implementation of rfc authors()

Does not handle RfcAuthor instances where person is None yet.

* refactor: authors -> bulk_authors for URL consistency

rudimatz committed:
fix: add OpenApi param definition doc_id (#9962)

* fix: add OpenApi param definition doc_id

* add schema def for person_id

jennifer-richards committed:
feat: RFC publication API (#9975)

* feat: API to publish RFC (WIP)

Incomplete and in need of refactoring, but publishes an RFC.

* feat: group / formal_languages from draft

* feat: allow optional formal_languages via API

Could do the same with group, but not clear it would ever be used.

* feat: fill in overrides/updates

* feat: subseries membership

* fix: tolerate race to create related docs

* fix: wrap pub in a transaction

* feat: prevent re-publishing draft as RFC

* chore: remove stale code

* chore: remove debug

* feat: RFC file upload API (WIP)

Checkpointing progress before going further.

* feat: specify RFC, validate file exts

* feat: move uploaded files into place

* feat: add replace option

* fix: add rest of replace option

* feat: handle ad/group more consistently

* chore: remove inadvertent change

* chore: drop external_url, get note from draft

* refactor: clarify default value logic

* refactor: ID obsoletes/updates by number

* fix: handle draft-stream-editorial

jennifer-richards committed:
feat: split unknown vs already published draft error (#10011)

jennifer-richards committed:
feat: use RfcAuthor for red API (#10014)

jennifer-richards committed:
chore: merge main into feat/rpc-api

jennifer-richards committed:
refactor: combine redundant serializers

jennifer-richards committed:
feat: edit authors via RFC update API

jennifer-richards committed:
fix: remove all None authors, not just first (#10044)

jennifer-richards committed:
refactor: extract update_rfcauthors() method

jennifer-richards committed:
feat: EditedRfcAuthorsDocEvent

jennifer-richards committed:
chore: merge main into feat/rpc-api

jennifer-richards committed:
Merge branch 'feat/rpc-api' into rfcauthor-docevents

jennifer-richards committed:
refactor: reduce RfcAuthor instance churn

jennifer-richards committed:
feat: create RfcAuthor edit DocEvents

jennifer-richards committed:
feat: handle DocumentAuthor->RfcAuthor updates

jennifer-richards committed:
refactor: reduce code duplication

jennifer-richards committed:
fix: remove leftover import

jennifer-richards committed:
chore: relabel titlepage_name in DocEvent desc

jennifer-richards committed:
feat: transaction

jennifer-richards committed:
chore: make RfcAuthorViewset read-only

This can perhaps go away entirely

jennifer-richards committed:
style: undo accidental whitespace change

jennifer-richards committed:
Merge pull request #10052 from jennifer-richards/rfcauthor-docevents

feat: better author edit via Rfc update API

jennifer-richards committed:
chore: merge main into feat/rpc-api

jennifer-richards committed:
fix: actual RFC file extensions/locations + tests (#10131)

* chore: update list of rfc file exts

* refactor: _destination() helper

* fix: .notprepped.xml -> prerelease/ subdir

* refactor: better prefixed DRF Routers

* test: fix references API test

* test: test notify_rfc_published

* test: test upload_rfc_files

* chore: remove unused imports

* chore: remove unused imports

* chore: add a todo

jennifer-richards committed:
Merge pull request #10216 from ietf-tools/main

chore: merge main into feat/rpc-api

jennifer-richards committed:
fix: find pubreq event for editorial stream

jennifer-richards committed:
Merge pull request #10217 from jennifer-richards/editorial-rfced-event

fix: find pubreq event for editorial stream

jennifer-richards committed:
chore: remove obsolete rpc demo API

jennifer-richards committed:
Merge pull request #10218 from jennifer-richards/no-demo-api

chore: remove obsolete rpc demo API

jennifer-richards committed:
chore: clean up outdated comments

jennifer-richards committed:
feat: api key for red api (#10232)

jennifer-richards committed:
fix: avoid over-return (#10231)

* fix: avoid over-return

* chore: undo accidental commit

This is a separate bug fix on main; let it come in from there.

jennifer-richards committed:
refactor: authors() -> author_persons() (#10237)

* refactor: authors() -> author_persons()

* refactor: select_related() a couple more places

* refactor: update uses of Document.authors()

* chore: remove debug

* fix: typo

jennifer-richards committed:
fix: mypy lint and minor bugs; mypy->1.11.2 (#10249)

rudimatz committed:
fix: doc property authors needs refactoring (#10250)

* fix: doc property needs refactoring

* fix: set source for author fields

* chore: comment

* chore: fix random typo

---------

Co-authored-by: Jennifer Richards <jennifer@staff.ietf.org>

jennifer-richards committed:
Merge branch 'main' into feat/rpc-api

jennifer-richards committed:
chore: remove completed todo, add comment

jennifer-richards committed:
chore(dev): accept devtoken for red api

rjsparks commented:
The test failures are only the expected changes in coverage (since the denominator moved).
